android/fakedapp/src/main/java/com/solana/mobilewalletadapter/fakedapp/MainViewModel.kt

```diff
@@ -419,56 +373,35 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
-                    semConnectedOrFailed.release()
-                }
-
-                override fun onScenarioError() = semConnectedOrFailed.release()
-                override fun onScenarioComplete() = semConnectedOrFailed.release()
-                override fun onScenarioTeardownComplete() = semTerminated.release()
-            }
-
             val localAssociation = LocalAssociationScenario(
                 getApplication<Application>().mainLooper,
                 Scenario.DEFAULT_CLIENT_TIMEOUT_MS,
-                scenarioCallbacks,
                 uriPrefix
             )
             sender.startActivityForResult(localAssociation.createAssociationIntent())
 
-            localAssociation.start()
-            try {
-                withTimeout(ASSOCIATION_TIMEOUT_MS) {
-                    semConnectedOrFailed.acquire()
+            return@withLock withContext(Dispatchers.IO) {
+                val mobileWalletAdapterClient = try {
+                    @Suppress("BlockingMethodInNonBlockingContext") // running in Dispatchers.IO; blocking is appropriate
+                    localAssociation.start().get(ASSOCIATION_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                } catch (e: InterruptedException) {
+                    Log.w(TAG, "Interrupted while waiting for local association to be ready")
+                    return@withContext null
+                } catch (e: TimeoutException) {
+                    Log.e(TAG, "Timed out waiting for local association to be ready")
+                    return@withContext null
+                } catch (e: ExecutionException) {
+                    Log.e(TAG, "Failed establishing local association with wallet", e.cause)
+                    return@withContext null
                 }
-            } catch (e: TimeoutCancellationException) {
-                Log.e(TAG, "Timed out waiting for local association to be ready", e)
-                // Let garbage collection deal with cleanup; if we timed out starting, we might
-                // hang if we attempt to close.
-                return@withLock null
-            }
 
-            val result = mobileWalletAdapterClient?.let { client -> action(client) } ?: run {
-                Log.e(TAG, "Local association not ready; skip requested action")
-                null
-            }
+                val result = action(mobileWalletAdapterClient)
 
-            localAssociation.close()
-            try {
-                withTimeout(ASSOCIATION_TIMEOUT_MS) {
-                    semTerminated.acquire()
-                }
-            } catch (e: TimeoutCancellationException) {
-                Log.e(TAG, "Timed out waiting for local association to close", e)
-                return@withLock null
-            }
+                @Suppress("BlockingMethodInNonBlockingContext") // running in Dispatchers.IO; blocking is appropriate
+                localAssociation.close().get(ASSOCIATION_TIMEOUT_MS, TimeUnit.MILLISECONDS)
```

Comment:
I was still really bothered by having to suppress the blocking warning so I did some more searching. Apparently, `suspend` functions aren't actually supposed to block, which I actually did not think was the case. So I *_think_*, what can happen here is that the IO thread can be blocked while the future waits for the websocket comes back. I'm not sure what we think about that, but I went looking for the quote-unquote "right" answer. I'd hate to force all our library consumers to suppress that warning.

I found two things that might help us:
- According to [this](https://stackoverflow.com/questions/63372461/await-java-5-futures-in-kotlin-coroutines-without-blocking-the-thread) Stackoverflow article, there might be a way to wrap the callbacks from a future combining them with an IO dispatcher converted into an executor. I'm not sure what the `ApiFuture` class is, but I'm sure we can figure out how it maps to our code.
- The other option is to use `CompleteableFuture` which the Kotlin coroutines library has an extension for, to turn it into a coroutine async/await operation, which I learned about [here](https://elizarov.medium.com/futures-cancellation-and-coroutines-b5ce9c3ede3a). That pulls our client lib up to minSdk 24, but I personally don't have a problem with that. 

Any thoughts about what we might do? I feel like we should probably nail down the way to do this without suppression, but maybe initially it's unavoidable.

Reply:
Okay, one more option. According to [this](https://stackoverflow.com/questions/50793362/how-to-use-suspendcoroutine-to-turn-java-7-future-into-kotlin-suspending-functio) SO post written by Roman Elizarov, Mr coroutines himself, we may just be able to cast our `Future` instances to `CompletionStage` and then used the `await()` extension method for proper coroutine integration.

Reply:
Hmm, that appears contrary to [Kotlin docs](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-dispatchers/-i-o.html):

> ... designed for offloading blocking IO tasks to a shared pool of threads.

Re: 

> Apparently, suspend functions aren't actually supposed to block

I don't think this is quite accurate. From [an article](https://elizarov.medium.com/blocking-threads-suspending-coroutines-d33e11bf4761) by Roman Elizarov,

> Suspending functions add a new dimension to code design. It was blocking/non-blocking without coroutines and now there is also suspending/non-suspending on top of that. To make everybody’s life simpler we use the following convention: suspending functions do not block the caller thread.

Which, AIUI, is exactly the purpose of Dispatchers.IO; to give an appropriate context for blocking calls within a `suspend` context, that isn't the calling coroutine context.

Reply:
Also, there are a bunch of complaints about this inspection online. It uses an overly aggressive metric: anything that throws `IOException` in a suspend context. IMO, the problem here is the inspection, not the code (which I'd argue is correct).

Reply:
Agreed - convenience layers can entirely hide this detail, so no one else needs to suffer at the hands of the overly aggressive inspection rule 😁